Update P3527 after implementation experience. (#7)
mpark/wg21-papers: P3527R0.md

matching design, since that language design informs the library design in this
paper.

To match based on the currently-engaged value of a variant-like object `v` of
type `V`, the pattern matching language feature uses `std::variant_size<V>`,
`std::variant_alternative<V>`, `std::get<T>(v)`, and `v.index()`.  The
`std::get` overloads used are the ones that take a type template parameter,

> not the ones that take a `size_t` NTTP.

Reviewer: It is actually `get<I>` that is used. The usage is basically like:
```
switch (v.index()) {
  case Idx: {
    std::variant_alternative_t<Idx, V>& __temp = get<Idx>(v);
  // for Idx 0 ... std::variant_size_v<V> - 1
  }
}
```
I wouldn't say `std::get`, since it's an ADL-only `get` call.

Author: Yeah, that's what I meant to write.  I just flipped it the first time, and then repeated the error after that.

# Design

The minimal design is this.

- Add an `index()` member function to `std::expected`.

- Add specializations of `std::variant_size`, and `std::variant_alternative`
  that work with `std::expeted`.

- Add type-parameter overloads of `std::get` that work with `std::expected`.

By making these changes, `std::expected` will be compatible with `match`
expressions wherever `std::variant` is.

> ## Optionl additions

Reviewer: ```suggestion
## Optional additions
```

> The changes above make `std::expeted` very close to `std::variant` in many

Reviewer: ```suggestion
The changes above make `std::expected` very close to `std::variant` in many
```

ways, but there are a few ways in which the two differ.  Since the main design
above brings them into such near-alignment, it might be useful to add even
more compatability, so that the user can treat `std::variant` and
`std::expected` as equivalent in most generic code.  The optional changes are:

- Add a version of `std::holds_alternative()` that works with `std::expected`.

- Add NTTP overloads of `std::get` that work with `std::expected`.

- Add overloads of `std::get_if` that work with `std::expected`.

The costs of wording and implementing these additions is low, so they seem
like good additions for the sake of consistency.

## Not included: modifying `std::visit`

The list of optional additions above leaves out a pretty important API --
`std::visit`.  These are the reasons the authors chose not to include it.

1) `std::expected` is fixed at two alternatives.  No one would write
`std::visit` when they can simply write `if`, unless perhaps they were
handling variant-like types generically, in which case...

2) ..., the advent of pattern matching as a language feature means we can use

> that for generic handling of variant-like types.  None of has to write

Reviewer: ```suggestion
that for generic handling of variant-like types.  No one has to write
```
I'm guessing that's what you meant here.
